hkmc2/shared/src/main/scala/hkmc2/semantics/ImplicitResolver.scala

```diff
@@ -0,0 +1,108 @@
+package hkmc2.semantics
+
+import mlscript.utils.*, shorthands.*
+import hkmc2.utils.TraceLogger
+import hkmc2.Raise
+import hkmc2.semantics.Elaborator.State
+import hkmc2.syntax.Ins
+import hkmc2.ErrorReport
+import hkmc2.Message.MessageContext
+import hkmc2.syntax.Tree.TermDef
+import hkmc2.syntax.Fun
+
+case class CtxArgImpl(private val param: Param) extends CtxArg:
+  
+  var arg: Opt[Term] = N
+  override def term: Opt[Term] = arg
+  
+  def populate(term: Term): Unit = 
+    arg = S(term)
+    
+  override def toString: String = term match
+    case N => s"CtxArg(${param.sign})"
+    case S(term) => s"CtxArg($term: ${param.sign})"
+
+object ImplicitResolver:
+  
+  /*
+   * An "implicit" or "instance" context, as opposed to the one in Elaborator.
+   */
+  case class ICtx(parent: Opt[ICtx], env: Map[Term, ICtx.Instance]):
+    
+    def +(sign: Term, sym: Symbol): ICtx = copy(env = env + (sign -> ICtx.Instance(sym)))
+    
+    def get(sign: Term): Opt[ICtx.Instance] = env.get(sign)
+  
+  object ICtx:
+    final case class Instance(sym: Symbol)
+      
+    val empty: ICtx = ICtx(N, Map.empty)
+
+class ImplicitResolver(val tl: TraceLogger)
+(using val raise: Raise, val state: State):
+  import tl.*
+  import ImplicitResolver.ICtx
+  
+  def resolve(t: Term)(using ictx: ICtx): Unit = 
+    t match
+      case blk: Term.Blk => resolveBlk(blk)
+      case Term.Tup(elem) => elem.foreach(resolveElem(_))
+      case _ => t.subTerms.foreach(resolve(_))
+  
+  def resolveElem(elem: Elem)(using ictx: ICtx): Unit = 
+    elem match
+      case arg @ CtxArgImpl(param) => 
+        log(s"Resolving context argument: $param")
+        val sign = param.sign match
+          case N =>
+            raise(ErrorReport(msg"Missing signature for context argument ${param.toString}" -> N :: Nil))
+            Term.Error
+          case Some(sign) =>
+            sign
+        ictx.get(sign) match
+          case N =>
+            raise(ErrorReport(msg"Missing instance for context argument ${param.toString}" -> N :: Nil))
+          case S(ICtx.Instance(sym)) =>
+            log(s"Resolved context argument: $param -> $sym")
+            arg.populate(sym.ref())
+      case _ => 
+        elem.subTerms.foreach(resolve(_))
+  
+  def resolveBlk(blk: Term.Blk)(using ictx: ICtx): ICtx = 
+  trace[ICtx](s"Resolving block: $blk", * => s"~> $blk"):
```

Comment:
Wow what's that cursed lambda? And it should be
```suggestion
  trace[ICtx](s"Resolving block: $blk", * => s"~> ${*}"):
```

Reply:
Because terms are mutable in this resolution stage, the functions work by side effects to the `blk`. So here I print the `blk` at beginning and print it again after the whole function is done. The param `*` here is an `ICtx`, which I thought it didn't need to be printed? I forgot why did I choose `*` as the name, but now I no more agree with myself 😶‍🌫️

Reply:
Ok, sure. And normally we'd just use `_` here. I didn't even know `*` could be bound as a parameter name like this.